Add Adaptive Thermal Comfort Thermostat Control

xuanluo113 opened the pull request:
Pull request overview
---------------------
The adaptive comfort model has been adopted in EnergyPlus as one of the thermal comfort models of the People object, as well as one control strategy to operate windows in the Airflow Network model. The existing code will be reused as much as possible. We propose to enhance the existing ThermostatSetpoint objects to indicate the applicability of the adaptive comfort model with three choices: 
- **None**. The adaptive comfort model is not applicable; 
- **AdaptiveASH55**. The central line of the ASHRAE Standard 55-2010 adaptive comfort model will be used as the zone operative temperature setpoint; and 
- **AdaptiveCEN15251**. The central line of the CEN Standard 15251-2007 adaptive comfort model will be used as the zone operative temperature setpoint. 

When the adaptive comfort model is selected, the thermostat setpoint temperature schedule will be overwritten with the calculated operative temperature based on the central line of the comfort model defined in ASHRAE 55-2010 or CEN 15251-2007. Such calculations have been implemented in EnergyPlus already. The ASHRAE adaptive comfort model is only applicable when the running average outdoor air temperature for the past 30 days is between 10.0 and 33.5°C; while the CEN 15251-2007 adaptive comfort model is only applicable when the running average outdoor air temperature for the past 7 days is between 10.0 and 30.0°C. 

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [x] Code style (parentheses padding, variable names)
 - [x] Functional code review (it has to work!)
 - [x] If defect, results of running current develop vs this branch should exhibit the fix
 - [x] CI status: all green or justified
 - [x] Performance: CI Linux results include performance check -- verify this
 - [x] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add rules to spreadsheet
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Documentation changes in place
 - [ ] Changed docs build successfully
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

rongpengzhang committed:
NFP

hongtz68 committed:
Design Document

NFP is updated to address Mike Witte's comments.

hongtz68 committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into 126735005-AdaptiveThermostat

hongtz68 committed:
Addressed Jason's comments

xuanluo113 committed:
First test

xuanluo113 committed:
Revert "First test"

This reverts commit 7f0e7158fe6a1a7b8f76c1270955f3fdec72a454.

xuanluo113 committed:
Cleaned up

xuanluo113 committed:
Clean up

xuanluo113 committed:
Revert "Clean up"

This reverts commit 17110f1892fd630e3112aeb6e9ea8678e4ad5998.

xuanluo113 committed:
Clean up

xuanluo113 committed:
Add design day environment

xuanluo113 committed:
add comment

xuanluo113 committed:
Merge branch 'develop' into 126735005-AdaptiveThermostat

xuanluo113 committed:
Addressed compiling err

xuanluo113 committed:
Addressed compiling err

xuanluo113 committed:
address warnings

xuanluo113 committed:
change cmake and idd

xuanluo113 committed:
style

lgu1234 commented on src/EnergyPlus/ZoneTempPredictorCorrector.cc:
```diff
@@ -5664,6 +5952,82 @@ namespace ZoneTempPredictorCorrector {
+			// Adjust run period cooling set point 
+			switch ( AdaptiveComfortModelTypeIndex ) {
+				case ASH55_CENTRAL:
+					ZoneAirSetPoint = AdapComfortDailySetPointSchedule.ThermalComfortAdaptiveASH55_Central( DayOfYear );
```
@XuanLuoLBL @hongtz68 I would like to have clarification. The output of AdapComfortDailySetPointSchedule.ThermalComfortAdaptiveASH55_Central is operative temperature based on ASHRAE 55 figure. My understanding is that ZoneAirSetPoint is dry air temperature. The radiative fraction is needed to convert the operative temperature to zone dry air temperature. Is my understanding correct?

Here is code extracted from AdjustAirSetPointsforOpTempCntrl:

		ZoneAirSetPoint = ( ZoneAirSetPoint - thisMRTFraction * thisMRT ) / ( 1.0 - thisMRTFraction );

xuanluo113 replied:
That is correct. In run time simulation, we overwrite the operative temperature first, and the dry air temperature based on radiative fraction is calculated afterwards,

lgu1234 commented on src/EnergyPlus/ZoneTempPredictorCorrector.cc:
```diff
@@ -1601,6 +1627,26 @@ namespace ZoneTempPredictorCorrector {
+						if ( TempControlledZone( TempControlledZoneNum ).OperativeTempControl ) {
+							if ( NumAlphas == 4 ) {
+								int adaptiveComfortModelTypeIndex = FindItem( cAlphaArgs( 4 ), AdaptiveComfortModelTypes, AdaptiveComfortModelTypes.isize() );
+								if ( adaptiveComfortModelTypeIndex != ADAP_NONE ) {
```
If adaptiveComfortModelTypeIndex == ADAP_NONE, do you provide additional message? For example, if it is wrong spelling, do you set ErrorsFound = true?

xuanluo113 replied:
Yes. We used FindItem() and reports errors when input string is not found in the type list. Though the checking order was a little messy so we addressed this in the latest commit.

lgu1234 commented on src/EnergyPlus/ZoneTempPredictorCorrector.cc:
```diff
@@ -2964,6 +3245,13 @@ namespace ZoneTempPredictorCorrector {
+				if ( ( TempControlledZone( RelativeZoneNum ).AdaptiveComfortTempControl ) ) {
+					AdjustOperativeSetPointsforAdapComfort( RelativeZoneNum, ZoneThermostatSetPointHi( ActualZoneNum ) );
+				}
+
```
I prefer to provide an additional warning, when the adaptive comfort thermostat has lower cooling setpoint than one obtained from the original thermostat.

xuanluo113 replied:
That is a very good point, Tianzhen and I discussed it and decided to add a new field indicating whether to overwrite when calculated temperature is lower than the original thermostat. Please see the latest commit for detail.
This helps a lot and thank you.

xuanluo113 committed:
Addressed Gu's comments

1. add overwrite options
2. fixed error and warning message
3. Improve style

xuanluo113 committed:
Add unit test

Add unit test function
EnergyPlusFixture.ZoneTempPredictorCorrector_AdaptiveThermostat

xuanluo113 committed:
Add docs

EnergyArchmage commented on src/EnergyPlus/ZoneTempPredictorCorrector.hh:
```diff
@@ -226,6 +247,12 @@ namespace ZoneTempPredictorCorrector {
 	CalcZoneAirTempSetPoints();
 
 	void
+	CalculateMonthlyRunningAverageDryBulb( Array1D< Real64 > & runningAverageASH, Array1D< Real64 > & runningAverageCEN );
```
this arg syntax seem unusual to me, any higher c++ guru care to take note? why not the "&" in this and not the next one.

xuanluo113 replied:
This function aims to calculate the value of the **runningAverageASH** and **runningAverageCEN** , so the pointer sign is to call by reference. The next one only uses the value of the two arrays, and I'd assume it's better only to call by value so the array won't be messed up in this function. Not sure if necessary or appropriate here.

mbadams5 replied:
@XuanLuoLBL In the other function, it is passing the arrays by value, which means it will make a copy of those arrays every time. You should pass them by const reference if you want them to not change, but still not make a new copy every time.

```cpp
CalculateAdaptiveComfortSetPointSchl( Array1D< Real64 > const & runningAverageASH, Array1D< Real64 > const & runningAverageCEN )
```

xuanluo113 replied:
@EnergyArchmage @mbadams5 Thanks for the clarification. Lesson learned. I've changed it in the latest commit.

EnergyArchmage commented on tst/EnergyPlus/unit/ZoneTempPredictorCorrector.unit.cc:
```diff
@@ -629,3 +633,293 @@ TEST_F( EnergyPlusFixture, ZoneTempPredictorCorrector_CorrectZoneHumRatTest )
+		AdjustOperativeSetPointsforAdapComfort( DualZoneNum, ZoneAirSetPoint );
+		ASSERT_EQ( 26.0, ZoneAirSetPoint );// Tstat should show set point is not overwritten
+
+	}
```
need a newline here

xuanluo113 replied:
done

EnergyArchmage commented on idd/Energy+.idd.in:
```diff
@@ -31933,10 +31933,28 @@ ZoneControl:Thermostat:OperativeTemperature,
+  A4,  \field Adaptive Comfort Model Type
+       \type choice
+       \key None
+       \key AdaptiveASH55_CentralLine
```
The use of underscore "_" in key choices is not allowed per the naming convention "Field choice names shall only use alphanumeric characters"
https://github.com/NREL/EnergyPlus/blob/develop/doc/module-developer/src/input-concepts/data-dictionary-naming-conventions.tex#L21
please remove the underscore in these key choices and adjust the get input and i/o reference accordingly

xuanluo113 replied:
Done.

EnergyArchmage commented on src/EnergyPlus/ZoneTempPredictorCorrector.cc:
```diff
@@ -2931,6 +3271,13 @@ namespace ZoneTempPredictorCorrector {
+
+				//Added Jan 17 (X. Luo)
+				//Adjust operative temperature based on adaptive comfort model
+				if ( ( TempControlledZone( RelativeZoneNum ).AdaptiveComfortTempControl ) ) {
```
Why call this for single heating setpoint? doesn't it only apply to cooling

EnergyArchmage committed:
review edits for adpative op temp control, testfile corrections

EnergyArchmage commented:
Okay, I have review this in some detail. Committed some edits to I/O ref description and some code revisions.  Testfile was not setup to actually use new feature at all and was fixed (names did not match), but you should review.  It would be better if the testfile exercised both the single-zone and the zone-list-type tstat input schemes. 

The AlwaysOverwrite options seems of little use to me and behavior should always be as for OverwriteWhenEnergySaving.  Night setup schedules makes this happen all the time -- your example has 40C setpoint at night and uses way more energy in AlwaysOverwrite...   I am afraid that it also increases the risk that ZoneThermostatSetPointHi comes out lower than ZoneThermostatSetPointLo which is a very big problem for dual setpoint load predictions.

I think what is missing here is that there needs to be a new output variable for the intermediate value generated by the adaptive models.  The user has no way of knowing what the new thermostat setpoints are after being overwritten in AdjustOperativeSetPointsforAdapComfort.  The output for "Zone Thermostat Cooling Setpoint Temperature" is subsequently modified by radiative conditions and the op temp inversion.  Probably need to store the intermediate value in a new variable analogous to ZoneThermostatSetPointHi and setup on new output variable on that.

xuanluo113 commented:
@EnergyArchmage Thanks for your review. That's very detailed and helpful. I'm working on the revision and I'm just a little confused regarding: "Testfile was not setup to actually use new feature at all and was fixed (names did not match), but you should review. It would be better if the testfile exercised both the single-zone and the zone-list-type tstat input schemes."

I saw your edition on the idf file line 2578. See that's something I inherited from another testfile 5ZoneEndUses.idf (only adding a new field to the original object). And by adding the zone name "SPACE3-1" before the thermostat name, I guess it complies with the reference "If only a single thermostat/zone is desired, then the name to be put here is <zone name> <Thermostat Name> where the Thermostat name is the thermostat that referenced the set of zones." And the testing result shows it only overwrite the set point in SPACE3-1. So why doesn't it use the new feature here? Do I have any misunderstanding in the naming and aliasing?

Thanks so much for your effort.

EnergyArchmage commented:
sorry, I don't ever used that zone list shortcut input method and am not very familiar with it, maybe @mjwitte know how it is supposed to work.  I thought you had to use the entire zone list, I could be wrong.  I have never heard of prepending a zone name to access just one of the zones in the list.  Seems too tricky by half for an example, even if it does work.  I had checked zone 3 but  didn't see any impact.

mjwitte commented:
When an object references a ZoneList, it creates clones of the object that are named \<ZoneName\>+\<ObjectName that's using ZoneList\>.  So, in this case
```
  ZoneList,
    AllControlledZones,      !- Name
    SPACE1-1,                !- Zone 1 Name
    SPACE2-1,                !- Zone 2 Name
    SPACE3-1,                !- Zone 3 Name
    SPACE4-1,                !- Zone 4 Name
    SPACE5-1;                !- Zone 5 Name

  ZoneControl:Thermostat,
    AllControlledZones Thermostat,  !- Name
    AllControlledZones,      !- Zone or ZoneList Name
```
creates 5 thermostats named
```
SPACE1-1 AllControlledZones Thermostat
SPACE2-1 AllControlledZones Thermostat
SPACE3-1 AllControlledZones Thermostat
SPACE4-1 AllControlledZones Thermostat
SPACE5-1 AllControlledZones Thermostat
```
From that point forward, any object that references one of these thermostats for reporting or other reasons (like operative temperature or in this case adaptive control) must use the expanded name.

xuanluo113 committed:
Addressed reviewer's comments

1. Added output variable
2. Delete always overwrite option
3. style

xuanluo113 commented:
@mjwitte Thanks for your explanation.

@EnergyArchmage 
I agree and I've addressed your comments. The latest commit reports a new variable "Zone Adaptive Comfort Operative Temperature Set Point" and it's easier to see the change. Please see the updated testfile.

Also, we agreed it's safer to delete the field "AlwaysOverwrite". Now there's no option to overwrite the setback or anytime when the original setpoint is higher. The unit test and documentation are updated correspondingly.

We also fixed the style.

EnergyArchmage commented:
Okay, this PR is ready to be merged.  Thanks for addressing the comments!
